[ADP-3359] Simplify construction of API errors relating to recent eras.

jonathanknowles opened the pull request:
This PR:
- adds a new constant `recentEras :: Set AnyRecentEra`
    - this constant defines the complete set of **_recent eras_** that are supported for transaction construction.
- uses this constant to simplify the construction of the following API errors:
    - `ApiErrorNodeNotYetInRecentEra`
    - `ApiErrorUnsupportedEra`

## Issue

ADP-3359

Follow-on from #4595.

jonathanknowles committed:
Add `Ord` instance for `AnyRecentEra`.

jonathanknowles committed:
Add function `supportedRecentEras`.

jonathanknowles committed:
Add function `supportedRecentEras` (to `Api.Types`).

jonathanknowles committed:
Use `Set` for supported eras in `ApiErrorNodeNotYetInRecentEra`.

jonathanknowles committed:
Use explicit imports when importing from `Cardano.Write.Tx`.

jonathanknowles committed:
Simplify construction of `ApiErrorNodeNotYetInRecentEra`.

jonathanknowles committed:
Simplify construction of `ApiErrorUnsupportedEra`.

jonathanknowles committed:
Remove redundant `$`.

jonathanknowles committed:
Further simplify construction of `ApiErrorUnsupportedEra`.

jonathanknowles committed:
Setify `supported_recent_eras` within `ApiError.json`.

This commit adjusts the golden file for `ApiError.json` so that
`supported_recent_eras` fields:

    - do not include duplicate eras
    - list eras in alphabetic order

This is necessary, as the Haskell type for this field has changed from a
list to a `Set`, which changes the behaviour of the generator for the
`ApiErrorNodeNotYetInRecentEra` data type.

Anviking commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs:
```diff
@@ -463,13 +471,21 @@ instance Bounded AnyRecentEra where
 
+-- | The complete set of supported recent eras.
+--
+supportedRecentEras :: Set AnyRecentEra
```
> **supported**RecentEras

What would an unsupported recent era be?

It seems like it's saying the same things twice. I.e. could it be more suitable to name it `allRecentEras` or  `supportedErasForTxConstruction` (the latter more for the api, not here) ?

jonathanknowles replied:
@Anviking good point, thanks for this feedback!

I've adjusted this in 18a34f0d5cf05488499032f368a24cf5c5ed2e5b.

jonathanknowles commented on lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx.hs:
```diff
@@ -463,13 +471,21 @@ instance Bounded AnyRecentEra where
     maxBound = AnyRecentEra RecentEraConway
 
+instance Ord AnyRecentEra where
+    compare = compare `on` fromEnum
```
This instance is needed in order to have a `Set AnyRecentEra`.

paolino reviewed (approved):
:+1:  (I think the `api` is the wrong place for computations, but this is not relevant for the refactoring)

jonathanknowles committed:
Rename `supportedRecentEras` to `allRecentEras`.

In response to review feedback:

https://github.com/cardano-foundation/cardano-wallet/pull/4603#discussion_r1609807012

paweljakubas reviewed (approved):
:+1:
